src/solve/slg/implementation/mod.rs

```diff
@@ -233,7 +233,176 @@ impl context::Environment<SlgContext> for Arc<Environment> {
     }
 }
 
-impl context::Substitution<SlgContext> for Substitution {}
+impl context::Substitution<SlgContext> for Substitution {
+    fn may_invalidate(&mut self, subst: &Substitution) -> bool {
+        self.parameters
+            .iter()
+            .zip(&subst.parameters)
+            .any(|(p1, p2)| MayInvalidate.aggregate_parameters(p1, p2))
+    }
+}
+
+// This is a struct in case we need to add state at any point like in AntiUnifier
+struct MayInvalidate;
+
+impl MayInvalidate {
+    fn aggregate_parameters(&mut self, p1: &Parameter, p2: &Parameter) -> bool {
+        match (p1, p2) {
+            (ParameterKind::Ty(ty1), ParameterKind::Ty(ty2)) => {
+                self.aggregate_tys(ty1, ty2)
+            }
+            (ParameterKind::Lifetime(l1), ParameterKind::Lifetime(l2)) => {
+                self.aggregate_lifetimes(l1, l2)
+            }
+            (ParameterKind::Ty(_), _) | (ParameterKind::Lifetime(_), _) => {
+                panic!("mismatched parameter kinds: p1={:?} p2={:?}", p1, p2)
+            }
+        }
+    }
+
+    // Returns true if the two types could be unequal.
+    fn aggregate_tys(&mut self, ty0: &Ty, ty1: &Ty) -> bool {
+        match (ty0, ty1) {
+            (Ty::Var(_), Ty::Var(_)) => false,
+
+            // Aggregating universally-quantified types seems hard according to Niko. ;)
+            (Ty::ForAll(_), Ty::ForAll(_)) => true,
+
+            (Ty::Apply(apply1), Ty::Apply(apply2)) => {
+                self.aggregate_application_tys(apply1, apply2)
+            }
+
+            (Ty::Projection(apply1), Ty::Projection(apply2)) => {
+                self.aggregate_projection_tys(apply1, apply2)
+            }
+
+            (Ty::UnselectedProjection(apply1), Ty::UnselectedProjection(apply2)) => {
+                self.aggregate_unselected_projection_tys(apply1, apply2)
+            }
+
+            (Ty::Var(_), ty) | (ty, Ty::Var(_)) => {
+                !self.recurse_ty(ty)
+            }
+
+            (Ty::ForAll(_), _)
+                | (Ty::Apply(_), _)
+                | (Ty::Projection(_), _)
+                | (Ty::UnselectedProjection(_), _) => true,
```

Comment:
PS, run rustfmt, I *think* this is not the style it produces?

Reply:
Will do! I don't normally use it, so I had just copied existing style (this format was done in the AntiUnifier).

Reply:
I don't think it's indented quite as much there; but yeah I prefr to run rustfmt...